keps/sig-storage/0031-20181029-VolumeSubpathEnvExpansion-apichange.md

```diff
@@ -0,0 +1,351 @@
+---
+kep-number: 0
+title: VolumeSubpathEnvExpansion
+authors:
+  - "@kevtaylor"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+  - sig-architecture
+reviewers:
+  - TBD
+  - "@msau42"
+  - "@thockin"
+  - "@liggitt"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-10-29
+last-updated: 2018-10-29
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+VolumeSubpathEnvExpansion API change
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Current workarounds - k8s &lt;=1.9.3](#current-workarounds---k8s-193)
+      * [Workarounds - k8s &gt;1.9.3](#workarounds---k8s-193)
+      * [Alternatives - using subPath directly](#alternatives---using-subpath-directly)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Alternatives - Using subPathFrom](#alternatives---using-subpathfrom)
+
+## Summary
+
+Legacy systems create all manner of log files and these are not easily streamed into stdout
+
+Files that are written to a host file path need to be uniquely partitioned
+
+If 2 or more pods run on the same host writing the same log file names to the same volume, they will clash
+
+Using the `subPath` is a neat option but because the `subPath` is "hardcoded" ie. `subPath: mySubPath` it does not enforce uniqueness
+
+To alleviate this, the `subPath` should be able to be configured from an environment variable as `subPath: $(MY_VARIABLE)` 
+
+The workaround to this issue is to use symbolic links or relative symbolic links but these introduce sidecar init containers
+and a messy configuration overhead to try to create upfront folders with unique names - examples of this complexity are detailed below
+
+## Motivation
+
+The initial alpha feature was implemented to allow unique addressing of subpaths on a host
+This cannot currently be achieved with the downwardAPI and requires complex workarounds
+
+The workarounds became more difficult after 1.9.3 when symbolic links were removed from initContainers
+
+### Goals
+
+To reduce excessive boiler-plate workarounds and remove the need for complex initContainers
+
+### Non-Goals
+
+Full template implementation for subPaths
+
+## Proposal
+
+The api change proposed is to create a Mutually Exclusive Field separate from the `subPath`
+called `subPathExpr`
+
+The subpath code which expands environment variables from the API would (under this proposal) change from
```

Comment:
Worth commenting that the expansion support is alpha today, and should not proceed because of compat reasons.